### ISSUE TITLE
[Feature Request] Support compilation to WebAssembly

### ISSUE THREAD

**Issue by redradist** [cython/cython#3698]
It would be nice to be able to compile Python to WebAssembly as AssemblyScript does https://github.com/AssemblyScript/assemblyscript

**Comment by scoder** [comment]
While nice, this isn't really something that needs to be built into Cython. It's more about setting up a build environment to pass the generated C file through, e.g. as part of [Pyodide](https://github.com/iodide-project/pyodide). I heard that a couple of people might be working on something like that already.

**Comment by redradist** [comment]
@scoder 
> 
> While nice, this isn't really something that needs to be built into Cython. It's more about setting up a build environment to pass the generated C file through, e.g. as part of [Pyodide](https://github.com/iodide-project/pyodide). I heard that a couple of people might be working on something like that already.

But Pyodide compiles whole infrastructure, it means that underneath it still interpreted
But I am taking about compiling strict sub-set of Python (typed Python) to WebAssembly as AssemblyScript did ...
If they can to do this stuff, we definitely also able to do the same ...

**Comment by rth** [comment]
You could try to Cythonize a simple cdef function, extract the generated C function and run emscripten on it to get WASM. As long as the generated C makes no references to the Python C API (or other external libraries outside of libc) it probably should work. I'm not sure it's possible to generate C functions that make no reference to the Python C API with Cython. Even if it is, the inability to use any Python objects will significantly limit what one can do with it.

I imagine you would need something that converts Python code to C or C++ without requiring libpython/CPython in the end. Maybe pythran could do it, since it apparently has a [C++ run time](https://pythran.readthedocs.io/en/latest/DEVGUIDE.html#c-runtime), not sure, but in general it's a somewhat orthogonal issue to building for web assembly.

**Comment by stonebig** [comment]
I see David Brochard tweet somewhat telling it's solved.
Is there a missing detail ?

![image](https://user-images.githubusercontent.com/4312421/222894478-4100d2e8-5cbd-4eee-9fc6-bc72c9ae9b50.png)

**Comment by da-woods** [comment]
We haven't done anything specific to make it work, but also we didn't really expect to. Cython's job is largely to generate the C code and let other tools take it from them.

With that said, it might be good to have a bit of documentation on how to get it to work... I certainly don't know

**Comment by rth** [comment]
Yeah, you still need to compile the C code generated by Cython anyway as far as I know. Either via [`pyodide build`](https://pyodide.org/en/stable/development/building-and-testing-packages.html#build-the-wasm-emscripten-wheel) to create a wheel or via [emscripten-forge](https://github.com/emscripten-forge/recipes).   So as long as the package is setup to build extensions with cython ( via `pyproject.toml` or `setup.py`) it should just work with one of those build systems targeting WebAssembly. 

> But I am taking about compiling strict sub-set of Python (typed Python) to WebAssembly as AssemblyScript did

If you really don't want to use Python interpreter, that would mean you cannot use the standard library, and Python packages since none of them would be in this strict subset of Python. So that would be of rather limited use. For instance, AssemblyScript re-implements the [JS standard library](https://github.com/AssemblyScript/assemblyscript/tree/main/std) , which would be a huge task for Python.

**Comment by da-woods** [comment]
> > But I am taking about compiling strict sub-set of Python (typed Python) to WebAssembly as AssemblyScript did
> 
> If you really don't want to use the Python interpreter, that would mean you cannot use the standard library, and Python packages since none of them would be in this strict subset of Python.

For all other uses Cython absolutely requires the Python interpreter. (Ok, for some incredibly simple code you can get away without, but that's luck and not something we officially support). So I don't think we'd even consider trying to compile to interpreter-less web assembly
